### ISSUE TITLE
Mismatched units in MathML causes segfault in Generator

### ISSUE THREAD

**Issue by kerimoyle** [cellml/libcellml#459]
I'm getting a segmentation fault in the generator for a model which is valid according to the validator, and has no errors from `generator.processModel()`.  Perhaps something to do with how equations are removed inside generateEquationCode?

Here's the file with my code: 
[generator_KRM.cpp.zip](https://github.com/cellml/libcellml/files/3850843/generator_KRM.cpp.zip)

No guarantees that I haven't screwed something up in setting up the model, but even if I have, whatever it is should be caught either by the validator or generator checks to return a useful error rather than segfaulting, if possible ... @agarny any ideas?

**Comment by kerimoyle** [comment]
So after some more digging the segfault comes from (or at least goes away when I fix ...) a mismatch of units within a MathML string.  Where's the best place for this to be checked? The `Validator` or the `Generator`?

**Comment by agarny** [comment]
I have just had a quick look at your code and, unless I missed something, I believe there is a problem with it. In step 1b, you have `equation2` which is defined as (by the way, you might want to respect the way we write MathML code in our C++ files, i.e. don't use `\` to continue a string over several lines):

```
<apply>
    <eq/>
    <ci>Na_conductance</ci>
    <apply>
        <times/>
        <apply>
            <power/>
            <ci>m</ci>
            <apply>
                <times/>
                <cn cellml:units="dimensionless">3</cn>
                <ci>h</ci>
            </apply>
        </apply>
    </apply>
</apply>
```

Now, we can see that the top `times` has only one argument, and this is the reason you are getting a segmentation fault (since the generator assumes that your model is valid). This can easily be confirmed by modifying the equation to read:

```
<apply>
    <eq/>
    <ci>Na_conductance</ci>
    <apply>
        <times/>
        <apply>
            <power/>
            <ci>m</ci>
            <apply>
                <times/>
                <cn cellml:units="dimensionless">3</cn>
                <ci>h</ci>
            </apply>
        </apply>
    </apply>
    <ci>m</ci>    <--- Added to confirm that the problem was a missing operand
</apply>
```

**Comment by kerimoyle** [comment]
Thanks @agarny .  In any case, I think that something that passes the validator shouldn't cause a segfault.  And really, I think that a user shouldn't be able to cause a segfault _anywhere_, but especially not in something which has passed the validator.  So the cause of the error (and it is possible to make it go away by changing the mathml, as in my second comment) is not really the purpose of this issue - the real purpose it to make sure that a user can't cause a segfault without knowing what's gone wrong (if at all!).  Either we need to check more _usage_ errors (cf _specification_ errors only) in the validator, or we need to catch more errors in the generator.  Or perhaps even both ...

**Comment by agarny** [comment]
That's the thing, I don't think your mdoel should have passed validation in the first place. Your MathML code is syntactically valid, but not semantically valid. As you know, once validation is complete, our code generator will first make sure that a model is valid and, if it is, then will go ahead with code generation. So, I would expect that check with regards to the number of operands for a given operator to be done in the validation step.

**Comment by kerimoyle** [comment]
This is the trouble I have... if the validator only validates against the spec, then it's currently doing its job.  If we want it to do more than that then there are a huge number of other places it could be checking things!  I think that all our functions should be failsafe - if the generator needs semantically correct maths but the validator doesn't, then the generator should check it?  Perhaps this is a time where we need the `Issue` class for warnings from the validator? What do others think? @hsorby @nickerso

**Comment by nickerso** [comment]
I think we need to be careful. There should be a validator that validates as per the rules of the CellML specification. Beyond that, there are more application-specific types of validation that can (and should) be supported by libCellML (or some service sitting on top of libCellML). I think in this particular case that is what #462 should be looking at. Dimensional checks within equations is yet another type of validation. I would like to see the validator class remain as a validation against the CellML specification and other types of validation layered on top of that - which will also, I'm pretty sure, start to necessitate the need for something along the lines of the `Issue` class.

I agree that we want to make sure that our code should be failsafe. I do think the generator should be able to operate by assuming it only ever works with models that have been checked to ensure they are valid for that operation, but this entails more than checking that a model is valid according to just the CellML spec.

**Comment by agarny** [comment]
Closing in favour of issue #688.
